File: Java/PathSumII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<>();
        path(root, sum, new ArrayList<Integer>(), res);
        return res;
    }

    private void path(TreeNode root, int sum, List<Integer> vals, List<List<Integer>> res) {
        if(root == null) return;
        vals.add(root.val);
        if(root.left == null && root.right == null && root.val == sum) {
            res.add(new ArrayList<>(vals));
        }
        
        path(root.left, sum - root.val, vals, res);
        path(root.right, sum - root.val, vals, res);

        vals.remove(vals.size() - 1);
    }
}

File: Go/path_sum_II.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) [][]int {
    var res [][]int
    var elem []int

    path(root, sum, &elem, &res)

    return res
}

func path(root *TreeNode, sum int, elem *[]int, res *[][]int) {
    if root == nil {
        return
    }
    *elem = append(*elem, root.Val)
    if root.Left == nil && root.Right == nil && sum == root.Val {
        r := append([]int{}, *elem...)
        *res = append(*res, r)
    }

    path(root.Left, sum - root.Val, elem, res)
    path(root.Right, sum - root.Val, elem, res)

    *elem = (*elem)[:len(*elem) - 1]
}